### PR TITLE
[Pytorch][Inductor] Add device-generic DRAM bandwidth estimates (#196398)

### PR DESCRIPTION
**Pull request by excalibur68**
Summary:

Add a device-aware `get_device_dram_gbps(device)` API backed by the shared `DeviceInfo` registry so runtime estimation no longer assumes CUDA. Thread the FX node device through AOTAutograd transfer-time estimation, preserve the legacy `get_gpu_dram_gbps()` behavior, and return `None` when bandwidth information is unavailable.

Test Plan:
Automated unit tests cover:
- registered device information and current-device selection
- unavailable and unknown device behavior
- legacy CUDA/XPU fallback behavior
- propagation of tensor device information into transfer-time estimation

Backend integration tests cover runtime device-information registration.

Relevant lint checks pass.

Reviewed By: IvanKobzarev

Differential Revision: D119241426

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/utils.py [hunk 1]
```diff
@@ -3745,14 +3745,99 @@ def get_device_tflops(dtype: torch.dtype) -> float:
...
+    ds_bw = datasheet_dram_bw_gbs(device_name) if device_name is not None else None
+    if ds_bw is not None:
+        return ds_bw
+
+    if device.type in ("cuda", "xpu"):
+        from triton.testing import get_dram_gbps
+
+        return get_dram_gbps(device.index)
+
+    log.warning(
```
Comment (blaine-rister) [inline torch/_inductor/utils.py:3813]: I'm wondering if this `inf` fallback would end up causing other bugs later on. What's the intended use case where we'd want to fall back? 

Would it be better to raise an exception in this function? That way we can quickly spot and fix the error instead of progressing with an incorrect DRAM estimate.

File: torch/_inductor/utils.py [hunk 2]
```diff
@@ -3745,14 +3745,99 @@ def get_device_tflops(dtype: torch.dtype) -> float:
...
+        name = properties.get("name", properties.get("arch"))
+    else:
+        name = getattr(properties, "name", None)
+    return name if isinstance(name, str) else None
+
+
+_device_dram_gbps_providers: dict[str, Callable[[torch.device], float | None]] = {}
+
+
+def register_device_dram_gbps_provider(
```
Comment (blaine-rister) [inline torch/_inductor/utils.py:3773]: While this is a very generic mechanism, it also introduces more global state where each backend needs to register its own provider. I'm wondering if this is overkill for such a niche function. It might be better to register a more generic class which stores other HW info as well. (For example, something which could also handle the tflops estimate added in your other PR.)

It looks like there's already a "datasheet" mechanism where each device can register its performance metrics. Would it be possible for MTIA to reuse that instead of registering a new function? If the concern is exposing internal HW stats to OSS, maybe we could add our datasheet to the table inside of fbcode, when the closed-source Inductor backend is registered.

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/analysis/device_info.py#L28

**Review by blaine-rister — COMMENTED** [review]
Left some comments about whether MTIA needs a new hook, or would the existing `DeviceInfo` mechanism work for this use case.

File: torch/_inductor/utils.py (hunk 2 above)
Comment (excalibur68) [reply torch/_inductor/utils.py:3773]: I think the major concern is we don't want to expose the internal HW stats to OSS at this moment. 
therefore we did that in fbcode, please see the diff part. 
This register provides such a bridge and could be resued by other accelerators.

Comment (blaine-rister) [reply torch/_inductor/utils.py:3773]: In that case, would it be possible to add MTIA's device info to this dict? That would let us unify the hook/registration mechanism between in-tree and out-of-tree backends. https://github.com/pytorch/pytorch/blob/main/torch/_inductor/analysis/device_info.py#L35

If we want to create a hook which officially declares this as an extension point, one way would be to have the function could add the `DeviceInfo` to this dict internally. Inductor follows a similar style for some other extension points like lowerings, codegen backends, etc.

Something like this could work.
```
from torch._inductor.analysis import device_info

def register_device_info(device: str, info: DeviceInfo):
    device_info._device_mapping[device] = info
```

File: torch/_inductor/utils.py (hunk 1 above)
Comment (blaine-rister) [reply torch/_inductor/utils.py:3813]: On closer inspection, it seems like the existing implementation returns `None` if the device isn't found. Perhaps we should stick to that. https://github.com/pytorch/pytorch/blob/main/torch/_inductor/analysis/device_info.py#L356

File: torch/_functorch/_aot_autograd/streams.py [hunk 3]
```diff
@@ -64,10 +64,18 @@ def map_value(x: Any) -> Any:
         for t in flat_outs
         if isinstance(t, torch.Tensor) and t.dtype in _FLOAT_TYPES
     }
+    device = next(
+        (
+            value.device
+            for value in [*flat_outs, *flat_args_kwargs]
```
Comment (mlazos) [inline torch/_functorch/_aot_autograd/streams.py:70]: _Reviewed by me, this may be an issue._

_AI-generated analysis (Codex):_

> **Device lookup ignores AOT metadata**
> 
> AOT nodes store tensors in meta["val"], but these inputs are derived through meta["value"]. The lookup therefore finds no Tensor and queries the current accelerator instead of the graph device. The test masks this by fabricating meta["value"].

File: torch/_inductor/utils.py (hunk 2 above)
Comment (excalibur68) [reply torch/_inductor/utils.py:3773]: yeah I can have a try

File: torch/utils/_runtime_estimation.py [hunk 4]
```diff
@@ -166,16 +166,20 @@ def get_transfer_time(flat_args_kwargs, flat_outs, gpu_type=None) -> float:  # t
...
+            the device has no registered bandwidth estimate.
     """
     if gpu_type is not None:
         gpu_memory_bandwidth = datasheet_dram_bw_gbs(gpu_type)
         if gpu_memory_bandwidth is None:
             raise ValueError(f"gpu_type {gpu_type!r} not found in datasheet")
     else:
-        gpu_memory_bandwidth = get_gpu_dram_gbps()
+        gpu_memory_bandwidth = get_device_dram_gbps(device)
+        if gpu_memory_bandwidth is None:
```
Comment (blaine-rister) [inline torch/utils/_runtime_estimation.py:181]: What's the motivation for returning 0 here vs. None?

File: test/inductor/test_utils.py [hunk 5]
```diff
@@ -60,6 +69,203 @@
 from torch.utils._sympy.functions import Identity
 
 
+class TestDeviceDramBandwidth(TestCase):
+    def test_get_device_info_key_uses_mapping_arch_then_name(self):
```
Comment (blaine-rister) [inline test/inductor/test_utils.py:105]: These tests currently need lot of boilerplate to mock various functions for the new device. There are a couple ways we might be able to reduce this.
 - Can we call `register_interface_for_device` to register a basic device interface? That way we wouldn't need to mock these functions; we could test the real one.
 - Can we refactor the device setup/registration to the class' `setUp` method? That way it will automatically be called at the start of each test.

File: test/inductor/test_utils.py [hunk 6]
```diff
@@ -392,6 +598,57 @@ def test_load_template_invalid_utf8_names_the_file(self):
 
 
 class TestRuntimeEstimation(TestCase):
+    def test_roofline_estimate_gets_device_from_tuple_metadata(self):
+        from torch._functorch._aot_autograd.streams import get_roofline_estimate
+
+        graph = torch.fx.Graph()
```
Comment (blaine-rister) [inline test/inductor/test_utils.py]: Related to this [comment](https://github.com/pytorch/pytorch/pull/196398/changes#r4010422417), is it possible to trace the graph with `torch.export` as opposed to constructing the FX graph directly? That way we could check that the realistic metadata is used.

File: test/inductor/test_utils.py [hunk 7]
```diff
@@ -60,6 +69,203 @@
...
+                    return_value=device_interface,
+                ),
+            ):
+                device_interface.Worker.get_device_properties.return_value = properties
+                self.assertEqual(
+                    _get_device_info_key(torch.device("mtia:0")), expected
+                )
+
+    def test_get_device_dram_gbps_uses_registered_device_info(self):
+        _get_device_dram_gbps.cache_clear()
```
Comment (blaine-rister) [inline test/inductor/test_utils.py]: Can we move these cache clearing ops into the class' `tearDown` method? That would avoid some duplication and let us remove the try/finally's in the tests.

**Review by IvanKobzarev — APPROVED** [review]

File: torch/_functorch/_aot_autograd/streams.py (hunk 3 above)
Comment (excalibur68) [reply torch/_functorch/_aot_autograd/streams.py:70]: Good catch. AOTAutograd uses meta["val"]; the manually constructed test was masking that mismatch. let me fix this

File: torch/_inductor/utils.py (hunk 1 above)
Comment (excalibur68) [reply torch/_inductor/utils.py:3795]: makes sense, let me make them consistent

File: torch/utils/_runtime_estimation.py (hunk 4 above)
Comment (excalibur68) [reply torch/utils/_runtime_estimation.py:181]: regarding this my understanding is: 
1. get_device_dram_gbps() is a lookup API, return None is reasonable since it mean unavailable 
2. but get_transfer_time() here is a calculation, returning None would raise TypeError. 
therefore return 0.0 preserves the old behavior which ignore the unavailable memory estimate and use compute time.

File: test/inductor/test_utils.py (hunk 5 above)
Comment (excalibur68) [reply test/inductor/test_utils.py:105]: yes changed accordingly

File: test/inductor/test_utils.py (hunk 7 above)
Comment (excalibur68) [reply test/inductor/test_utils.py]: sure

**Comment by excalibur68** [comment]
Two AOTI failures are stale generated-code expectations under TORCHINDUCTOR_LITE_MODE.
Two Dynamo failures are the same stale expected-failure marker reporting an “unexpected success.”
the failures are not related to this change

**Review by blaine-rister — APPROVED** [review]
LGTM! Thanks for the quick responses to feedback.

The only improvement I would suggest is potentially registering a device interface in `setUp`, which might be more maintainable than the mocks which are currently used.  See here: https://github.com/pytorch/pytorch/pull/196398/changes#r4029022741

This need not block merging the PR though. It's not a big deal, just a tiny nit.

File: test/inductor/test_utils.py [hunk 8]
```diff
@@ -60,6 +69,168 @@
...
+                raise cls.error
+            return cls.properties[device]
+
+
+class TestDeviceDramBandwidth(TestCase):
+    def setUp(self):
+        super().setUp()
+        self.device = torch.device("privateuseone:0")
+        _TestDeviceInterface.Worker.properties = {}
+        _TestDeviceInterface.Worker.error = None
```
Comment (excalibur68) [inline test/inductor/test_utils.py:89]: Tests side also configure Worker.properties and exercise the real get_interface_for_device() path. 
so no more direct device-interface mocks
